src/lib/yang/data.lua

```diff
@@ -1492,6 +1494,94 @@ function print_data_for_schema(schema, data, file, is_config)
    return data_printer_from_schema(schema, is_config)(data, file)
 end
 
+function consistency_checker_from_grammar(grammar)
+   -- Returns the corresponding data path for a grammar node.
+   local function data_path (root, node)
+      local ret = {}
+      local function visit (root)
+         for k,v in pairs(root) do
+            if k == 'members' then
+               visit(v)
+            elseif v.members then
+               table.insert(ret, k)
+               visit(v.members)
+            elseif root == node then
+               table.insert(ret, k)
+               return
+            end
```

Comment:
Does this work for choices?  I am a little wary here, like maybe we should switch exhaustively on the grammar kind rather than seeing what properties the node has.

I wonder, what about:

```lua
local function visit_leaf_paths(root)
   local function visit(path, node)
      if node.type == 'struct' then
         for k,v in pairs(node.members) do visit(path..'/'..k, v) end
      elseif node.type == 'leaf' then
         coroutine.yield(path, node)
      elseif node.type == 'array' then
         -- pass
      elseif node.type == 'table' then
         for k,v in pairs(node.keys) do visit(path..'/'..k, v) end
         for k,v in pairs(node.values) do visit(path..'/'..k, v) end
      elseif node.type == 'choice' then
         for _,choice in pairs(node.choices) do
            for k,v in pairs(choice) do visit(path..'/'..k, v) end
         end
      else error('unexpected kind', node.kind) end
   end
   return coroutine.wrap(function() visit('', root)), true
end
```

Then you can do `for path, node in visit_leaf_paths(root) do ... end`.  Dunno.  WDYT?  Tree traversal is a classic use for coroutines!